File: src/main/java/org/labun/jooq/codegen/config/Defaults.java
package org.labun.jooq.codegen.config;

import org.labun.jooq.codegen.task.SchemaCodeGenerationTask;
import org.labun.jooq.codegen.task.TableCodeGenerationTask;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * @author Konstantin Labun
 */
public class Defaults {

    private Defaults() {
    }

    public static Configuration configuration() {
        return new Configuration()
                .codeGeneration(new ArrayList<>(Arrays.asList(
                        CodeGenerationConfigs.COLUMNS_GENERATION_CONFIG,
                        CodeGenerationConfigs.RECORD_GENERATION_CONFIG,
                        CodeGenerationConfigs.TABLE_GENERATION_CONFIG,
                        CodeGenerationConfigs.SCHEMA_GENERATION_CONFIG,
                        CodeGenerationConfigs.REPOSITORY_GENERATION_CONFIG
                )));
    }

    public static class GeneratorsNames {
        public static final String SCHEMA = "DEFAULT:SCHEMA";
        public static final String TABLE = "DEFAULT:TABLE";
        public static final String RECORD = "DEFAULT:RECORD";
        public static final String COLUMNS = "DEFAULT:COLUMNS";
        public static final String REPOSITORY = "DEFAULT:REPOSITORY";

        private GeneratorsNames() {
        }
    }

    public static class CodeGenerationConfigs {
        public static final CodeGenerationConfig TABLE_GENERATION_CONFIG = tableGenerationConfig();
        public static final CodeGenerationConfig RECORD_GENERATION_CONFIG = recordGenerationConfig();
        public static final CodeGenerationConfig COLUMNS_GENERATION_CONFIG = columnsGenerationConfig();
        public static final CodeGenerationConfig SCHEMA_GENERATION_CONFIG = schemaGenerationConfig();
        public static final CodeGenerationConfig REPOSITORY_GENERATION_CONFIG = repositoryGenerationConfig();

        private CodeGenerationConfigs() {
        }

        private static CodeGenerationConfig schemaGenerationConfig() {
            return new CodeGenerationConfig()
                    .codeGenerationTask(SchemaCodeGenerationTask.class.getCanonicalName())
                    .generatorName(GeneratorsNames.SCHEMA)
                    .packageName("org.labun.jooq.codegen")
                    .className(new NameConfig())
                    .template("templates/java/schema.vm");
        }

        private static CodeGenerationConfig columnsGenerationConfig() {
            return new CodeGenerationConfig()
                    .codeGenerationTask(TableCodeGenerationTask.class.getCanonicalName())
                    .generatorName(GeneratorsNames.COLUMNS)
                    .packageName("org.labun.jooq.codegen.column")
                    .className(new NameConfig().postfix("Columns"))
                    .template("templates/java/columns.vm")
                    .javaTimeDates(true);
        }

        private static CodeGenerationConfig recordGenerationConfig() {
            return new CodeGenerationConfig()
                    .codeGenerationTask(TableCodeGenerationTask.class.getCanonicalName())
                    .generatorName(GeneratorsNames.RECORD)
                    .packageName("org.labun.jooq.codegen.record")
                    .className(new NameConfig().postfix("Record"))
                    .template("templates/java/record.vm")
                    .javaTimeDates(true);
        }

        private static CodeGenerationConfig tableGenerationConfig() {
            return new CodeGenerationConfig()
                    .codeGenerationTask(TableCodeGenerationTask.class.getCanonicalName())
                    .generatorName(GeneratorsNames.TABLE)
                    .packageName("org.labun.jooq.codegen.table")
                    .className(new NameConfig().postfix("Table"))
                    .template("templates/java/table.vm")
                    .javaTimeDates(true);
        }

        private static CodeGenerationConfig repositoryGenerationConfig() {
            return new CodeGenerationConfig()
                    .codeGenerationTask(TableCodeGenerationTask.class.getCanonicalName())
                    .generatorName(GeneratorsNames.REPOSITORY)
                    .packageName("org.labun.jooq.codegen.repository")
                    .className(new NameConfig().postfix("RepositoryBase"))
                    .template("templates/java/repository.vm")
                    .javaTimeDates(true);
        }

        public static List<CodeGenerationConfig> all() {
            return Arrays.asList(
                    tableGenerationConfig(),
                    recordGenerationConfig(),
                    columnsGenerationConfig(),
                    schemaGenerationConfig(),
                    repositoryGenerationConfig());
        }
    }

    public static class TemplateVariables {
        public static final String SCHEMA = "schema";
        public static final String TABLE = "table";
        public static final String ID = "id";
        public static final String COLUMNS = "columns";
        public static final String PACKAGE = "package";
        public static final String CLASS_NAME = "className";
        public static final String COLUMNS_CLASS_NAME = "columnsClassName";
        public static final String TABLE_CLASS_NAME = "tableClassName";
        public static final String RECORD_CLASS_NAME = "recordClassName";
        public static final String UNIQ_KEYS = "uniqKeys";

        private TemplateVariables() {
        }
    }
}

File: src/main/kotlin/org/labun/jooq/generator/Defaults.kt
package org.labun.jooq.generator

import org.labun.jooq.generator.config.Configuration
import org.labun.jooq.generator.config.NameConfig
import org.labun.jooq.generator.config.SubGeneratorConfig
import org.labun.jooq.generator.generation.GeneratorsNames
import org.labun.jooq.generator.task.CatalogSubGenerator
import org.labun.jooq.generator.task.SchemaSubGenerator
import org.labun.jooq.generator.task.TableSubGenerator

/**
 * @author Konstantin Labun
 */
object Defaults {

  @JvmStatic fun defaultConfiguration(): Configuration {
    return Configuration(
        subGenerators = listOf(
            CodeGenerationConfigs.CATALOG_GENERATION_CONFIG,
            CodeGenerationConfigs.SCHEMA_GENERATION_CONFIG,
            CodeGenerationConfigs.SEQUENCES_GENERATION_CONFIG,
            CodeGenerationConfigs.INDEXES_GENERATION_CONFIG,
            CodeGenerationConfigs.TABLE_GENERATION_CONFIG,
            CodeGenerationConfigs.RECORD_GENERATION_CONFIG,
            CodeGenerationConfigs.COLUMNS_GENERATION_CONFIG,
            CodeGenerationConfigs.KEYS_GENERATION_CONFIG,
            CodeGenerationConfigs.REPOSITORY_GENERATION_CONFIG
        ))
  }

  object CodeGenerationConfigs {

    val CATALOG_GENERATION_CONFIG = catalogGenerationConfig()
    val SCHEMA_GENERATION_CONFIG = schemaGenerationConfig()
    val SEQUENCES_GENERATION_CONFIG = sequenceGenerationConfig()
    val INDEXES_GENERATION_CONFIG = indexesGenerationConfig()
    val TABLE_GENERATION_CONFIG = tableGenerationConfig()
    val RECORD_GENERATION_CONFIG = recordGenerationConfig()
    val COLUMNS_GENERATION_CONFIG = columnsGenerationConfig()
    val KEYS_GENERATION_CONFIG = keysGenerationConfig()
    val REPOSITORY_GENERATION_CONFIG = repositoryGenerationConfig()

    private fun catalogGenerationConfig() = SubGeneratorConfig(
        generatorName = GeneratorsNames.CATALOG,
        subGenerator = CatalogSubGenerator::class.java.canonicalName,
        packageName = "org.labun.jooq.subgenerators",
        className = NameConfig(postfix = "Catalog"),
        template = "templates/java/catalog.vm"
    )

    private fun schemaGenerationConfig() =
        SubGeneratorConfig(
            generatorName = GeneratorsNames.SCHEMA,
            subGenerator = SchemaSubGenerator::class.java.canonicalName,
            packageName = "org.labun.jooq.subgenerators",
            className = NameConfig(),
            template = "templates/java/schema.vm"
        )

    private fun sequenceGenerationConfig() =
        SubGeneratorConfig(
            generatorName = GeneratorsNames.SCHEMA_SEQUENCES,
            subGenerator = SchemaSubGenerator::class.java.canonicalName,
            packageName = "org.labun.jooq.subgenerators",
            className = NameConfig(postfix = "Sequences"),
            template = "templates/java/sequences.vm"
        )

    private fun indexesGenerationConfig() =
        SubGeneratorConfig(
            generatorName = GeneratorsNames.SCHEMA_INDEXES,
            subGenerator = SchemaSubGenerator::class.java.canonicalName,
            packageName = "org.labun.jooq.subgenerators",
            className = NameConfig(postfix = "Indexes"),
            template = "templates/java/indexes.vm"
        )

    private fun tableGenerationConfig() =
        SubGeneratorConfig(
            subGenerator = TableSubGenerator::class.java.canonicalName,
            generatorName = GeneratorsNames.TABLE,
            packageName = "org.labun.jooq.subgenerators.table",
            className = NameConfig(postfix = "Table"),
            template = "templates/java/table.vm",
            javaTimeDates = true
        )

    private fun recordGenerationConfig(): SubGeneratorConfig {
      val subGeneratorConfig = SubGeneratorConfig(
          subGenerator = TableSubGenerator::class.java.canonicalName,
          generatorName = GeneratorsNames.TABLE_RECORD,
          packageName = "org.labun.jooq.subgenerators.record",
          className = NameConfig(postfix = "Record"),
          template = "templates/java/record.vm",
          javaTimeDates = true
      )
      return subGeneratorConfig
    }

    private fun columnsGenerationConfig() =
        SubGeneratorConfig(
            subGenerator = TableSubGenerator::class.java.canonicalName,
            generatorName = GeneratorsNames.TABLE_COLUMNS,
            packageName = "org.labun.jooq.subgenerators.column",
            className = NameConfig(postfix = "Columns"),
            template = "templates/java/columns.vm",
            javaTimeDates = true
        )

    private fun keysGenerationConfig() =
        SubGeneratorConfig(
            subGenerator = SchemaSubGenerator::class.java.canonicalName,
            generatorName = GeneratorsNames.SCHEMA_KEYS,
            packageName = "org.labun.jooq.subgenerators",
            className = NameConfig(postfix = "Keys"),
            template = "templates/java/keys.vm"
        )

    private fun repositoryGenerationConfig(): SubGeneratorConfig {
      return SubGeneratorConfig(
          subGenerator = TableSubGenerator::class.java.canonicalName,
          generatorName = GeneratorsNames.REPOSITORY,
          packageName = "org.labun.jooq.subgenerators.repository",
          className = NameConfig(postfix = "RepositoryBase"),
          template = "templates/java/repository.vm",
          javaTimeDates = true
      )
    }

    @JvmStatic fun all(): List<SubGeneratorConfig> = listOf(
        schemaGenerationConfig(),
        sequenceGenerationConfig(),
        indexesGenerationConfig(),
        keysGenerationConfig(),
        tableGenerationConfig(),
        recordGenerationConfig(),
        columnsGenerationConfig(),
        repositoryGenerationConfig(),
        catalogGenerationConfig())

    @JvmStatic fun base(): List<SubGeneratorConfig> = listOf(
        schemaGenerationConfig(),
        sequenceGenerationConfig(),
        indexesGenerationConfig(),
        keysGenerationConfig(),
        tableGenerationConfig(),
        recordGenerationConfig(),
        columnsGenerationConfig(),
        catalogGenerationConfig())
  }

}